File: Jim/src/sk/fiit/jim/garbage/plan/Plan.java
package sk.fiit.jim.garbage.plan;

import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;
import sk.fiit.jim.agent.AgentInfo;
import sk.fiit.jim.agent.models.EnvironmentModel;
import sk.fiit.jim.agent.models.WorldModel;
import sk.fiit.jim.agent.skills.HighSkill;
import sk.fiit.robocup.library.geometry.Angles;

/**
 * Rewrited Plan from ruby
 * 
 * @author Benkovic
 */

/* Overall HIGHSKILL NEEDED:
 * Pick a LOWSKILL
 * FormationHelper
 */

public class Plan extends Thread implements IPlan {

	public List<HighSkill> Highskill_Queue = new ArrayList<HighSkill>();
	public boolean turned = false;
	//TODO encapsutalion!!!
	public boolean beamed = false;
	public Angles straight_range1 = new Angles(0.0, 15.0);
	public Angles straight_range2 = new Angles(345.0, 360.0);
	

	public void run() {

		/*
		 * METHOD CAN BE USE ! turned_to_goal (); see_ball(); ball_front();
		 * ball_back(); is_ball_mine(); straight(); near_ball(); ball_unseen();
		 */

		control();
	}

	public void control() {
		
		if (Highskill_Queue.isEmpty() == true) {
			replan();
		} else {
			try {
				Highskill_Queue.get(0).execute();
				if (Highskill_Queue.get(0).isEnded()) {
					Highskill_Queue.remove(0);
				}
			} catch (Exception ex) {
				Logger.getLogger(Plan.class.getName()).log(Level.SEVERE, null,
						ex);

			}

		}

	}

	public void replan() {

	}

	public boolean turned_to_goal() {
		if (straight_range1.include(AgentInfo.getInstance().kickTarget().getPhi())
				|| straight_range2.include(AgentInfo.getInstance().kickTarget().getPhi())) {
			return true;
		}
		return false;
	}
	//TODO presunut tieto metody do prislusnych modelov

	public boolean ball_front() {
		return AgentInfo.getInstance().getWhereIsBall() == "front";
	}

	public boolean ball_back() {
		return AgentInfo.getInstance().getWhereIsBall() == "back";
	}

	public boolean is_ball_mine() {
		return AgentInfo.getInstance().isBallMine();
	}

	public boolean straight() {
		if (straight_range1.include(AgentInfo.getInstance().kickTarget().getPhi())
				|| straight_range2.include(AgentInfo.getInstance().ballControlPosition().getPhi())) {
			return true;
		}
		return false;
	}

	public boolean near_ball() {
		return AgentInfo.getInstance().nearBall();
	}

	public double ball_unseen() {
		return EnvironmentModel.SIMULATION_TIME
				- WorldModel.getInstance().getBall().getLastTimeSeen();
	}

}

File: Old_xml_and_scripts/Jim/plan/old-predUcesanim/plan.rb
########	plan.rb	 ########
#	@author: Androids		#
#############################
import Java::sk.fiit.jim.agent.models.EnvironmentModel

require 'singleton'
require 'thread'

lock = Mutex.new
lock.synchronize do 

class Plan
  include Singleton
  
  def change_skill move_name 
     @queue.clear
     @queue << LowSkill.new(move_name)
  end
 
  @@straight_range1 = Walk.angle_range(0.0, 15.0)
  @@straight_range2 = Walk.angle_range(345.0, 360.0)
  
  def initialize
    puts "Initializing!"
    @agentModel = Java::sk.fiit.jim.agent.models.AgentModel.getInstance
  	@agentInfo = Java::sk.fiit.jim.agent.AgentInfo.getInstance
  	@tacticalInfo = Java::sk.fiit.jim.agent.models.TacticalInfo.getInstance
  	@Angles = Java::sk.fiit.robocup.library.geometry.Angles
    @formationPosition    = @tacticalInfo.getFormPositionGlobal()
    @plan = []
    @queue = []
    @names = []
    @turned = false
    @beamed = false
  end
  
  def control
  	if @plan.empty? or @plan[0].isEnded()
  		replan
    end
    
	  unless @plan.empty?
	    @plan[0].execute
    end
  end
  
  def replan
  end
  
  def reset
  	@plan.clear
  	@beamed = false
  end
  
  def turned_to_goal?
  	@@straight_range1.include?( @kick_target_phi) or @@straight_range2.include?( @kick_target_phi)
  end

  def see_ball?
   	ball.notSeenLongTime() < 5
  end
  
  def ball_front?
  	@agentInfo.getWhereIsBall.eql? 'front'	
  end
  
  def ball_back?
  	@agentInfo.getWhereIsBall.eql? 'back'
  end
  
  def is_ball_mine?
  	return @agentInfo.isBallMine
  end
 
  def straight?
    @@straight_range1.include?(@target_position_phi) or @@straight_range2.include?(@target_position_phi)
  end
  
  def near_ball?
  	return @agentInfo.getNearBall
  end

  def closest_to_ball?
    return true
  end
  
  def ball_unseen
    EnvironmentModel.SIMULATION_TIME - WorldModel.getInstance.getBall.getLastTimeSeen()
  end

  #def fillQueue

	#return
  #end
  
end

  #------------override Ruby's global logging => log PLANNING instead--------------
  
  def log message = ""
    Log.log LogType::PLANNING, message
  end

  def debug message = ""
    Log.debug LogType::PLANNING, message
  end

  def error message = ""
    Log.error LogType::PLANNING, message
  end
end

#reset on 2nd load => causing replan
Plan.instance.reset